gps_ubxm8_i2c/gps.cpp

```diff
@@ -0,0 +1,117 @@
+#include "gps.h"
+
+#include "ubxHelpers.h"
+#include "ubxMessages.h"
+#include "ubxPacket.h"
+
+#define I2C_BUFFER_SIZE 1024
+
+GPS::GPS() {
+    packetReader = UBXPacketReader();
+    state = GPS::State::REQUEST_NOT_SENT;
+}
+
+const GPS::State GPS::getState() { return state; }
+
+/**
+ * Sends a request for position data if none are currently pending.
+ * Then, checks if there is data available from the GPS. If there is none,
+ * it just returns a NO_DATA response. If there is data, it reads up to 1024 bytes (configurable at the top of gps.cpp)
+ * After reading, it will return one of these:
+ * - RECEIVE_IN_PROGRESS: it found a valid header for the data we're looking for (UBX packet) but hasn't read all of it yet.
+ * - NO_UBX_DATA: it successfully read data but there was no header for the data we're looking for (UBX packet)
+ * - POLL_JUST_FINISHED: it got data. Now you can read it with GPS::getSolution.
+ *
+ * There are also special return types for different types of error, that are hopefully verbose enough to explain themselves.
+ */
+const GPS::PollResult GPS::pollUpdate(I2C_HandleTypeDef* i2c) {
+    if (state == GPS::State::REQUEST_NOT_SENT) {
+        uint8_t message[4] = {0x01, 0x07, 0x00, 0x00};
+        sendUBX(message, 4, &hi2c3);
+        state = GPS::State::POLLING_RESPONSE;
+    }
+
+    if (state == GPS::State::POLLING_RESPONSE) {
+        uint8_t lenBytes[2];
+        uint16_t dataLen = 0;
+        uint8_t buffer[I2C_BUFFER_SIZE];
+
+        HAL_StatusTypeDef dataLenReadStatus = HAL_I2C_Mem_Read(i2c, 0x42 << 1, 0xFD, 1, lenBytes, 2, 100);
+        if (dataLenReadStatus != HAL_OK) {
+            return GPS::PollResult::DATA_LEN_POLL_FAILED;
+        }
+
+        dataLen = lenBytes[0] << 8 | lenBytes[1];
+        if (dataLen == 0) {
+            return GPS::PollResult::NO_DATA;
+        }
+
+        if (dataLen > I2C_BUFFER_SIZE) {
```

Comment:
I think if the gps module has more than `I2C_BUFFER_SIZE` bytes of data available, we should still parse it during one execution of `pollUpdate()` and not wait for the next cycle

Reply:
I don't think we should do this because this risks blocking for a long time. Maybe a compromise could be reading at most a fixed multiple of I2C_BUFFER_SIZE like 5 times? Lmk if that sounds good.

Reply:
I see, I didn't consider that. Let's just leave it as is.